documentation/modules/auxiliary/scanner/scada/modbus_banner_grabbing.md

```diff
@@ -0,0 +1,87 @@
+## Vulnerable Application
+
+This module will perform Banner grabbing in devices that uses the Modbus protocol.
+The module will send a payload with the function code 43 which is used to read device identification.
+For more technical information, you can refer to this link: https://en.wikipedia.org/wiki/Modbus#Available_function/command_codes.
+
+By default the service is running on port 502, so any device with this port open could be a potential target.
+A search in Shodan with the Dork `port:502` it's an easy way to find targets.
+Also you can filter the results by 'product'.
+
+## Verification Steps
+  1. Do: ```use auxiliary/scanner/scada/modbus_banner_grabbing```
+  2. Do: ```set RHOST <IP>```
+  3. Do: ```set UNIT_ID <ID>```
+  4. Do: ```run```
+
+The response to be received may contain the following objects:
+
+```vendor name, product code, min. max. revision, vendor url, product name, model name, etc.```
+
+If the ```UNIT_ID``` value is not the correct one or the response contains an error you will receive one Modbus exception code message.
```

Comment:
Fixed the quote issue on this one, but there is still the issue of explaining how one would retrieve this value. @itsecurityco do you have any info on this?

Reply:
@gwillcox-r7 the Modbus/TCP messaging protocol requires that all nodes wishing to utilize messaging have a configurable IP + Unit ID identifier to uniquely determine the device address on the Modbus/TCP network. Valid Unit ID addresses range between 0 and 247 (with 255 usually reserved for communication directly with a gateway). Value is 0 when the target IP address has exactly one device.

So you can give a shot with 0 and if you don't receive the fingerprint you can try 1,2,3, etc.

Ref. https://modbus.org/docs/Modbus_Messaging_Implementation_Guide_V1_0b.pdf

```Unit Identifier
This field is used for routing purpose when addressing a device on a MODBUS+
or MODBUS serial line sub-network. In that case, the “Unit Identifier” carries the
MODBUS slave address of the remote device:
If the MODBUS server is connected to a MODBUS+ or MODBUS Serial Line
sub-network and addressed through a bridge or a gateway, the MODBUS Unit
identifier is necessary to identify the slave device connected on the subnetwork behind the bridge or the gateway. The destination IP address
identifies the bridge itself and the bridge uses the MODBUS Unit identifier to
forward the request to the right slave device.
The MODBUS slave device addresses on serial line are assigned from 1 to
247 (decimal). Address 0 is used as broadcast address.
On TCP/IP, the MODBUS server is addressed using its IP address; therefore, the
MODBUS Unit Identifier is useless. The value 0xFF has to be used.
When addressing a MODBUS server connected directly to a TCP/IP network,
it’s recommended not using a significant MODBUS slave address in the “Unit
Identifier” field. In the event of a re-allocation of the IP addresses within an
automated system and if a IP address previously assigned to a MODBUS
server is then assigned to a gateway, using a significant slave address may
cause trouble because of a bad routing by the gateway. Using a nonsignificant slave address, the gateway will simply discard the MODBUS PDU
with no trouble. 0xFF is recommended for the “Unit Identifier" as nonsignificant value.
Remark : The value 0 is also accepted to communicate directly to a
MODBUS/TCP device. ```

Reply:
Hmm so I am a little confused by this paragraph. Lets walk through it:

```
The MODBUS slave device addresses on serial line are assigned from 1 to
247 (decimal). Address 0 is used as broadcast address.
```

Okay so address 0 on a serial line is used as the broadcast address. We also see that address 1 to 247 are individual device IDs. However all this only applies to serial lines, which is not what this module is using. So lets read on.

```
On TCP/IP, the MODBUS server is addressed using its IP address; therefore, the
MODBUS Unit Identifier is useless. The value 0xFF has to be used.
```

So to me this is interesting. In tests we never had this be an issue, but according to the official documentation, when we connect via TCP/IP, the Unit Identifier should always be set to a value of 0xFF or 255. This suggests that we might not even need to brute force the ID, as its irrelevant here and is technically meant to always be 0xFF or some nonsignificant address.

```
Remark : The value 0 is also accepted to communicate directly to a
MODBUS/TCP device.
```

Okay so looks like based off of this the values 0xFF and 0x0 are acceptable values for direct communication via TCP/IP. So perhaps would it not be better to use one of these two values rather than trying to brute force all the IDs?

Reply:
I made some test with the following IP `MjEyLTE2Ni0yNDctMjA=`, `ODYtNjEtNzYtMzY=` and this is the results so far.

`MjEyLTE2Ni0yNDctMjA=` response when UNIT_ID is set to 255 : **Fingerprint**

`MjEyLTE2Ni0yNDctMjA=` response when UNIT_ID is set to 0 : **Fingerprint**

`MjEyLTE2Ni0yNDctMjA=` response when UNIT_ID is set to a random number 137 : **Fingerprint**

`ODYtNjEtNzYtMzY=` response when UNIT_ID is set to 255 : **Fingerprint**

`ODYtNjEtNzYtMzY=` response when UNIT_ID is set to 0 : **No data was received**

`ODYtNjEtNzYtMzY=` response when UNIT_ID is set to  a random number 137 : **Fingerprint**

With `MTkzLTc3LTk2LTE4Mw==` IP I got the same results as with `ODYtNjEtNzYtMzY=` . 

So I think `0xFF` (255) it's the value that we can setup into the payload and remove `UNIT_ID` from `register_options`.

Reply:
That sounds good to me, do you want me to make those changes so long for you? Otherwise happy for you to make the edits if you would prefer. Sounds like sticking to the `0xFF` would be the best option.

Reply:
Woops, just saw that you already addressed this with your last commits. Closing this issue as its now been resolved.